### PR TITLE
Allow searching article content in backend

### PR DESCRIPTION
**Pull request by smehrbrodt**
Pull Request for Issue #20072 .

### Summary of Changes

Introduced a 'content:' prefix to search the article content

### Testing Instructions

Go to Content->Articles in backend. Try searching for article content by prefixing your search with `content:`, e.g. `content:foo`.

### Expected result

Articles which have the string "foo" in their content should be found.

### Documentation Changes Required

Tooltip and maybe help needs to be updated.

### CODE REVIEW

**Comment by laoneo** [comment]
Searching in content should work without prefix. Nobody will know that you have to prefix it with "content:". I would leave that and just include to search in the article content.

**Comment by smehrbrodt** [comment]
> Searching in content should work without prefix. Nobody will know that you have to prefix it with "content:". I would leave that and just include to search in the article content.

Ok, have updated the patch accordingly.

**Comment by ggppdk** [comment]
This would be nice,
but it will have terrible performance on websites with a few thousand articles or more
even timeouts if you have 10,000+ articles

the clause `LIKE %word%` is OK on 
- a **varchar** 255 title column (or 400 for alias column) with a few backend users

but LIKE %word% on 2 longtext columns ?

And even if you use a real search index, and solve the performance issue
you will hit the UX problem
- users 90% - 95% of the time would search for words that they want to find in the title

so if you enable this always it would be better to priotirize results and return matched articles having the words in the title first and then matches in the content

or add a selector
Search in title
Search in title and content

**Comment by csthomas** [comment]
I suggest to replace prefixes with a new select box as @ggppdk mentioned.

Example:
```html
<select name="searchIn">
  <option value="title">Search in Title</option>
  <option value="content">Search in Content</option>
  <option value="id">Search by Id</option>
  ...
</select>
```

**Comment by ggppdk** [comment]
Yes a selector exactly like @csthomas suggested will be good UX and will give this possibility to small websites without destroying performance on large websites (as searching in content with LIKE will be optional and not default behaviour)

**Comment by brianteeman** [comment]
@laoneo at a minimum the search should only search the title by default. If it searches everything by default then it will effect performance. The prefix is the correct method. As is the ability to set the prefix in a select

**Comment by mbabker** [comment]
IMO if these quick filters are going to start searching content and not just titles it needs to be consistent across core.  I don't believe the intent of these filters was to ever be a full search similar to the frontend com_search and its plugins, and personally I'm not sure that we should make these filters behave as a full search, but if we're going to I'd suggest that the filters should most likely offer searching the same fields that the com_search plugins query at a minimum per component.

**Comment by smehrbrodt** [comment]
Ok then I will go back to the prefix.

Putting all available prefixes in a select box is a good idea, but out of scope for this patch.
That should be done separately, and consistently for all components.

**Comment by laoneo** [comment]
If you guys really want to have an explicit search in the full content, then I'm also in favor of @csthomas solution with a drop down which defines in what to search. We should stop adding hidden features which are not obvious and will never get documented.

**Comment by smehrbrodt** [comment]
> If you guys really want to have an explicit search in the full content, then I'm also in favor of @csthomas solution with a drop down which defines in what to search. We should stop adding hidden features which are not obvious and will never get documented.

I see no general problem with using prefixes for searching. I see it as a power user feature which I personally like very much (btw also available here on Github).

But creating a dropdown for this can still be done afterwards to make it more visible.

**Comment by Quy** [comment]
Update language string:
`COM_CONTENT_FILTER_SEARCH_DESC="Search in title and alias. Prefix with ID: or AUTHOR: or CONTENT: to search for an article ID, article author or article content."`

**Comment by laoneo** [comment]
The difference between Github and Joomla is that on Github it is properly documented https://help.github.com/articles/searching-issues-and-pull-requests/ which is not the case for Joomla. So, if this feature gets ever into Joomla, only the people participating in this pr, the guy who merged it and you do know it.

Don't get me wrong, I really like the idea, but it should be done properly.

**Comment by smehrbrodt** [comment]
> The difference between Github and Joomla is that on Github it is properly documented which is not the case for Joomla.

I updated the tooltip which now describes the 3 available prefixes: id, author and content.
Is there any other places where this needs to be documented?

**Comment by brianteeman** [comment]
if we stop using everything in joomla that isnt documented then there wont be much left ;)

File: administrator/language/en-GB/en-GB.com_content.ini [hunk 1]
```diff
@@ -118,7 +118,7 @@ COM_CONTENT_FIELDS_TYPE_MODAL_ARTICLE="Article"
 COM_CONTENT_FIELDSET_PUBLISHING="Publishing"
 COM_CONTENT_FIELDSET_RULES="Permissions"
 COM_CONTENT_FIELDSET_URLS_AND_IMAGES="Images and Links"
-COM_CONTENT_FILTER_SEARCH_DESC="Search in title and alias. Prefix with ID: or AUTHOR: to search for an article ID or article author."
+COM_CONTENT_FILTER_SEARCH_DESC="Search in title and alias. Prefix with ID: or AUTHOR: or CONTENT: to search for an article ID, article author or article content."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_content.ini]: can you change this please to 
+COM_CONTENT_FILTER_SEARCH_DESC="Search in title and alias. Prefix with ID: or AUTHOR: or CONTENT: to search for an article ID, article author or search in article content."

**Comment by tonypartridge** [comment]
This is what I was talking about the other day when you did the issue, without prefix we also face a user issue. Users are not used to searching full content this could have a mass negative effect on the community. 

Prefix is ok... but even I only found out about the ID prefix last week 😭.

Ideally we should leave as is, and add in additional filter options. Multi-select option I.e. title is selected by default, the. Allow description, custom fields, image etc.

**Comment by brianteeman** [comment]
restarted drone

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on f8d42f1c39d50a6b0e8294a9c29be0edc76b5f1f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20083">issues.joomla.org/tracker/joomla-cms/20083</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on f8d42f1c39d50a6b0e8294a9c29be0edc76b5f1f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20083">issues.joomla.org/tracker/joomla-cms/20083</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20083">issues.joomla.org/tracker/joomla-cms/20083</a>.</sub>

**Comment by infograf768** [comment]
Taking off RTC as this should also be implemented in the featured model as the change in the string is displayed there too
screen shot 2018-05-09 at 08 07 59

Please complete PR.

**Comment by infograf768** [comment]
Waiting for PR to be completed<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20083">issues.joomla.org/tracker/joomla-cms/20083</a>.</sub>

**Comment by smehrbrodt** [comment]
> Taking off RTC as this should also be implemented in the featured model as the change in the string is displayed there too

Good catch, fixed now!

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 234bf558948f0b683a1daadd0c9e900c8c158876<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20083">issues.joomla.org/tracker/joomla-cms/20083</a>.</sub>

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 234bf558948f0b683a1daadd0c9e900c8c158876<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20083">issues.joomla.org/tracker/joomla-cms/20083</a>.</sub>

**Comment by infograf768** [comment]
Back to RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20083">issues.joomla.org/tracker/joomla-cms/20083</a>.</sub>

**Pull request merged**